### DISCUSSION TITLE
Is there any way of saving and loading a whole model for loading without initialization?

### DISCUSSION THREAD

**Discussion by JLiu-1** [Lightning-AI/pytorch-lightning#6780] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I've already known that the most recommended way of saving and loading a model is saving and loading the state dict, such as:
```
model = MyLightningModule(hparams)
trainer.fit(model)
trainer.save_checkpoint("example.ckpt")
new_model = MyModel.load_from_checkpoint(checkpoint_path="example.ckpt")
```

However, this load method needs to initialize a model having the same structure and configuration as the saved one. In some circumstances, I may just need the model for inference, and would not like to first initialize the model (for convenience, or sometimes the information of the loaded model is missing). Therefore, I'd like to know that: 
1. Is there any way to fully save a model in PL which can get loaded even without knowing its structure?
2. If so, what's the problem it may bring when saving models in this way?

In pyTorch, the torch.save may meet my need, but I'm not sure how to use it in PL. For example, in the following codes:
```
model=Mymodel(......)
trainer=Trainer(......)
trainer.fit(model)
```
on which can I use the torch.save? Can I just use torch.save on the trainer object? If so, can this way works for my need? Thanks.

**Comment by awaelchli** [comment]
the problem with saving the whole model and it's "structure" is that it needs to also save all the source code basically. That's how the pickling works in Python. But not everything can be pickled always.

In your case, since you mention inference, I recommend using torchscript.
You can convert your LightningModule to torchscript and then save that to a file (and load it for inference):

Here is the method and also some example:
https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#to-torchscript

Also have a look at the PyTorch docs for torchscript
